File: src/main/scala/de/mineformers/allomancy/block/AllomanticMetalOre.scala
package de.mineformers.allomancy.block

import java.util

import de.mineformers.allomancy.block.properties.PropertyString
import net.minecraft.block.Block
import net.minecraft.block.material.Material
import net.minecraft.block.state.{BlockState, IBlockState}
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.{Item, ItemBlock, ItemStack}
import net.minecraft.util.EnumWorldBlockLayer
import net.minecraftforge.fml.relauncher.{SideOnly, Side}
import tessera._

/**
  * AllomanticMetalOre
  *
  * @author PaleoCrafter
  */
object AllomanticMetalOre extends Block(Material.rock) {
  lazy final val Names = Seq("copper", "zinc", "tin", "aluminium", "chromium")
  lazy final val Metal = new PropertyString("metal", Names: _*)

  setDefaultState(blockState.getBaseState.withProperty(Metal, Names.head))
  setUnlocalizedName("allomantic_metal_ore")
  setCreativeTab(CreativeTabs.tabBlock)

  def fromMetal(metal: String) = getDefaultState.withProperty(Metal, metal)

  @SideOnly(Side.CLIENT)
  override def getSubBlocks(itemIn: Item, tab: CreativeTabs, list: util.List[ItemStack]): Unit = {
    for (dmg <- Names.indices)
      list.add(new ItemStack(itemIn, 1, dmg))
  }

  @SideOnly(Side.CLIENT)
  override def getBlockLayer: EnumWorldBlockLayer = EnumWorldBlockLayer.CUTOUT

  override def damageDropped(state: IBlockState): Int = getMetaFromState(state)

  override def getMetaFromState(state: IBlockState): Int = Names.indexOf(state.getValue(Metal))

  override def getStateFromMeta(meta: Int): IBlockState =
    getDefaultState
      .withProperty(Metal, Names(meta max 0 min (Names.length - 1)))

  override def getHarvestTool(state: IBlockState): String = null

  override def createBlockState(): BlockState = createState(this, Metal)

  class OreItemBlock(block: Block) extends ItemBlock(block) {
    setHasSubtypes(true)

    override def getMetadata(damage: Int): Int = damage

    override def getUnlocalizedName(stack: ItemStack): String =
      s"tile.${Names(stack.getItemDamage max 0 min (Names.length - 1))}_ore"
  }

}

File: src/main/java/de/mineformers/allomancy/block/AllomanticMetalOre.java
package de.mineformers.allomancy.block;

import de.mineformers.allomancy.block.properties.PropertyString;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.BlockState;
import net.minecraft.block.state.IBlockState;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumWorldBlockLayer;
import net.minecraft.util.MathHelper;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.Arrays;
import java.util.List;

/**
 * AllomanticMetalOre
 *
 * @author PaleoCrafter
 */
public class AllomanticMetalOre extends Block
{
    public static final String[] NAMES = {
        "copper", "zinc", "tin", "aluminium", "chromium"
    };
    public static final PropertyString METAL = new PropertyString("metal", NAMES);

    public static int clampDamage(int value)
    {
        return MathHelper.clamp_int(value, 0, NAMES.length - 1);
    }

    public AllomanticMetalOre()
    {
        super(Material.rock);
        setDefaultState(blockState.getBaseState().withProperty(METAL, NAMES[0]));
        setUnlocalizedName("allomantic_metal_ore");
        setCreativeTab(CreativeTabs.tabBlock);
        setRegistryName("allomantic_metal_ore");
    }

    public IBlockState fromMetal(String metal)
    {
        return getDefaultState().withProperty(METAL, metal);
    }

    @Override
    public void getSubBlocks(Item item, CreativeTabs tab, List<ItemStack> list)
    {
        for (int dmg = 0; dmg < NAMES.length; dmg++)
        {
            list.add(new ItemStack(item, 1, dmg));
        }
    }

    @Override
    @SideOnly(Side.CLIENT)
    public EnumWorldBlockLayer getBlockLayer()
    {
        return EnumWorldBlockLayer.CUTOUT_MIPPED;
    }

    @Override
    public int damageDropped(IBlockState state)
    {
        return getMetaFromState(state);
    }

    @Override
    public IBlockState getStateFromMeta(int meta)
    {
        return getDefaultState().withProperty(METAL, NAMES[clampDamage(meta)]);
    }

    @Override
    public int getMetaFromState(IBlockState state)
    {
        return Arrays.asList(NAMES).indexOf(state.getValue(METAL));
    }

    @Override
    protected BlockState createBlockState()
    {
        return new BlockState(this, METAL);
    }

    public static class ItemRepresentation extends ItemBlock
    {
        public ItemRepresentation(Block block)
        {
            super(block);
            setHasSubtypes(true);
        }

        @Override
        public int getMetadata(int damage)
        {
            return damage;
        }

        @Override
        public String getUnlocalizedName(ItemStack stack)
        {
            return String.format("tile.%s_ore", NAMES[clampDamage(stack.getItemDamage())]);
        }
    }
}
